cassiopeia/tools/small_parsimony.py

```diff
@@ -0,0 +1,230 @@
+
+    Returns:
+        A new CassiopeiaTree if the copy is set to True, else None.
+
+    Raises:
+        A CassiopeiaTreeError if Fitch-Hartigan bottom-up has not been called
+        or if the state_key does not exist for a node.
+    """
+
+    # assign root
+    root = cassiopeia_tree.root if (root is None) else root
+
+    cassiopeia_tree = cassiopeia_tree.copy() if copy else cassiopeia_tree
+
+    for node in cassiopeia_tree.depth_first_traverse_nodes(
+        source=root, postorder=False
+    ):
+
+        if node == root:
+            root_states = cassiopeia_tree.get_attribute(root, state_key)
+            cassiopeia_tree.set_attribute(
+                root, label_key, np.random.choice(root_states)
+            )
+            continue
+
+        parent = cassiopeia_tree.parent(node)
+        parent_label = cassiopeia_tree.get_attribute(parent, label_key)
+        optimal_node_states = cassiopeia_tree.get_attribute(node, state_key)
+
+        if parent_label in optimal_node_states:
+            cassiopeia_tree.set_attribute(node, label_key, parent_label)
+
+        else:
+            cassiopeia_tree.set_attribute(
+                node, label_key, np.random.choice(optimal_node_states)
+            )
+
+    return cassiopeia_tree if copy else None
+
+
+def score_small_parsimony(
+    cassiopeia_tree: CassiopeiaTree,
+    meta_item: str,
+    root: Optional[str] = None,
+    infer_ancestral_states: bool = True,
+    label_key: Optional[str] = "label",
+) -> int:
+    """Computes the small-parsimony of the tree.
+
+    Using the meta data stored in the specified cell meta column, compute the
+    parsimony score of the tree.
+
+    Args:
+        cassiopeia_tree: CassiopeiaTree object with cell meta data.
+        meta_item: A column in the CassiopeiaTree cell meta corresponding to a
+            categorical variable.
+        root: Node to treat as the root.
+        infer_ancestral_states: Whether or not ancestral states must be inferred
+            (this will be False if `fitch_hartigan` has already been called on
+            the tree.)
+        label_key: If ancestral states have already been inferred, this key
+            indicates the name of the attribute they're stored in.
+
+    Returns:
+        The parsimony score.
+
+    Raises:
+        CassiopeiaError if label_key has not been populated.
+    """
+
+    cassiopeia_tree = cassiopeia_tree.copy()
+
+    if infer_ancestral_states:
+        fitch_hartigan(cassiopeia_tree, meta_item, root, label_key=label_key)
+    
+    parsimony = 0
+    for (parent, child) in cassiopeia_tree.depth_first_traverse_edges(source=root):
+        
+        try:
+            if cassiopeia_tree.get_attribute(parent, label_key) != cassiopeia_tree.get_attribute(child, label_key):
+                parsimony += 1
+        except CassiopeiaTreeError:
+            raise CassiopeiaError(f"{label_key} does not exist for a node, "
+                                "try running Fitch-Hartigan.")
```

Comment:
Would also be useful to include a statement about `infer_ancestral_states` parameter. Something like `or provide infer_ancestral_states=True`.